q42stats/src/main/java/com/q42/q42stats/library/pay/GooglePayManager.kt

```diff
@@ -0,0 +1,117 @@
+package com.q42.q42stats.library.pay
+
+import android.content.Context
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.wallet.IsReadyToPayRequest
+import com.google.android.gms.wallet.Wallet
+import com.google.android.gms.wallet.WalletConstants
+import com.q42.q42stats.library.Q42StatsLogger
+import com.q42.q42stats.library.util.suspendCoroutineWithTimeout
+import kotlinx.coroutines.TimeoutCancellationException
+import org.json.JSONArray
+import org.json.JSONObject
+import kotlin.coroutines.resume
+
+private const val CHECK_GOOGLE_PAY_TIMEOUT = 5000L
+
+class GooglePayManager(context: Context) {
+
+    private val paymentsClient =
+        Wallet.getPaymentsClient(
+            context,
+            Wallet.WalletOptions.Builder()
+                .setEnvironment(WalletConstants.ENVIRONMENT_PRODUCTION)
+                .build()
+        )
+
+    suspend fun getSupportedCardNetworks(cardNetworks: List<CardNetwork>): Map<CardNetwork, IsReadyToPayResponse> =
+        mutableMapOf<CardNetwork, IsReadyToPayResponse>().apply {
+            cardNetworks.forEach { cardNetwork ->
+                put(cardNetwork, checkGooglePaySupport(cardNetwork))
+            }
+        }
+
+    /**
+     * We could test all cards at once, but we test every CardNetwork separately, so that we can
+     * log the exact CardNetwork of the user.
+     */
+    private suspend fun checkGooglePaySupport(cardNetwork: CardNetwork): IsReadyToPayResponse =
+        try {
+
+            /**
+             * When there are no Play Services on the device, or when they are disabled, a bug in Google Wallet
+             * causes isReadyToPay to never return a response. This timeout makes sure we are
+             * not blocking the collection for too long when this happens.
+             */
+            suspendCoroutineWithTimeout(CHECK_GOOGLE_PAY_TIMEOUT) { cont ->
+                val request = createReadyToPayRequest(listOf(cardNetwork.value))
+                paymentsClient.isReadyToPay(request).addOnCompleteListener { completedTask ->
+                    cont.resume(
+                        try {
+                            when (completedTask.getResult(ApiException::class.java)) {
```

Comment:
Assuming ApiException is something that will be caught, but couldn't find the javadoc. Can you confirm?

https://developers.google.com/pay/api/android/guides/tutorial#isreadytopay

Reply:
This is the documentation of the method: https://developers.google.com/android/reference/com/google/android/gms/tasks/Task#throws_1
Example here: https://developers.google.com/pay/api/android/guides/tutorial#isreadytopay
In all Google examples (also for other tasks), they call it with ApiException and don't say why.

I presumed it's the exception all unknown/unexpected thrown exceptions will be wrapped in, but am not completely sure. To be sure I'm catching all exceptions in the line below. It's closed code so no way to read it